chain-ser/src/deser.rs

```diff
@@ -1,28 +1,140 @@
+    InvalidData(String),
+}
+
+impl std::error::Error for ReadError {}
+
+impl std::fmt::Display for ReadError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            ReadError::NotEnoughBytes(left, demanded) => write!(
+                f,
+                "NotEnoughBytes: demanded {} bytes but got {}",
+                demanded, left
+            ),
+            ReadError::UnconsumedData(len) => write!(f, "Unconsumed data: {} bytes left", len),
+            ReadError::SizeTooBig(e, limit) => write!(
+                f,
+                "Ask for number of elements {} above expected limit value: {}",
+                e, limit
+            ),
+            ReadError::StructureInvalid(s) => write!(f, "Structure invalid: {}", s),
+            ReadError::UnknownTag(t) => write!(f, "Unknown tag: {}", t),
+            ReadError::InvalidData(s) => write!(f, "Invalid data: {}", s),
+        }
+    }
+}
+
+impl From<std::io::Error> for ReadError {
+    fn from(err: std::io::Error) -> Self {
+        ReadError::InvalidData(err.to_string())
+    }
+}
+
 /// Define that an object can be written to a `Write` object.
 pub trait Serialize {
-    type Error: std::error::Error + From<std::io::Error>;
-
-    fn serialize<W: std::io::Write>(&self, writer: W) -> Result<(), Self::Error>;
+    fn serialize<W: std::io::Write>(&self, codec: &mut Codec<W>) -> Result<(), WriteError>;
 
     /// Convenience method to serialize into a byte vector.
-    fn serialize_as_vec(&self) -> Result<Vec<u8>, Self::Error> {
-        let mut data = vec![];
-        self.serialize(&mut data)?;
+    fn serialize_as_vec(&self) -> Result<Vec<u8>, WriteError> {
+        let mut data = Vec::new();
+        self.serialize(&mut Codec::new(&mut data))?;
         Ok(data)
     }
 }
 
-/// Define that an object can be read from a `Read` object.
+impl<T: Serialize> Serialize for &T {
+    fn serialize<W: std::io::Write>(&self, codec: &mut Codec<W>) -> Result<(), WriteError> {
+        (*self).serialize(codec)
+    }
+}
+
 pub trait Deserialize: Sized {
-    type Error: std::error::Error + From<std::io::Error> + Send + Sync + 'static;
+    fn deserialize<R: std::io::Read>(codec: &mut Codec<R>) -> Result<Self, ReadError>;
 
-    fn deserialize<R: std::io::BufRead>(reader: R) -> Result<Self, Self::Error>;
+    fn deserialize_validate<R: std::io::Read>(codec: &mut Codec<R>) -> Result<(), ReadError> {
+        Self::deserialize(codec).map(|_| ())
+    }
 }
 
-impl<T: Serialize> Serialize for &T {
-    type Error = T::Error;
+pub trait DeserializeFromSlice: Sized {
+    fn deserialize_from_slice(codec: &mut Codec<&[u8]>) -> Result<Self, ReadError>;
 
-    fn serialize<W: std::io::Write>(&self, writer: W) -> Result<(), T::Error> {
-        (**self).serialize(writer)
+    fn deserialize_validate_from_slice(codec: &mut Codec<&[u8]>) -> Result<(), ReadError> {
+        Self::deserialize_from_slice(codec).map(|_| ())
```

Comment:
Is there any specific need for that? It should be essentially the same as `deserialize_validate` since `&[u8]` is also `Read`.

Reply:
Ok I see we are actually using `get_slice` in certain places, that makes sense.

Reply:
Or can we just implement a specialized variant of `Deserialize`?